Implemented and tested bulk declaration of bad replicas using lfns, r…

ivmfnal opened the pull request:
Implemented and unit-tested:
  - bulk declaration of bad replicas using a file with LFNs and --rse --scope in UI, with corresponding changes in client, web, api, core
  - bulk quarantining replicas using a file with paths and --rse, with corresponding changes in client, web, api, core
<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->

faluchet committed:
Consistency checks: Stop relying on AGIS to get RSEs information, use core API instead - Fix #5226

ChristophAmes committed:
Recovery: Change log messages, fix sleep_time propagation, add passive_mode #5300

joeldierkes committed:
REST & API: Change the `dids` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

KarthikSundar2002 committed:
Rules: Makes the update-rule locked field insensitve #5356

fno2010 committed:
Add new option --replica-sort to rucio download

fno2010 committed:
Enable --replica-sort for aria2c impl

jamesp-epcc committed:
Protocols: remove legacy S3 protocols and libraries #5346

cserf committed:
Deletion : Heartbeat is never refreshed for the Dark Reaper : Closes #5374

ChristophAmes committed:
Recovery: Change argeparse option for passive mode #5300

cserf committed:
get_pfn_to_rse() doesn’t properly handle protocols with different ports : Closes #5180

cserf committed:
Core : Bug in delete_dids when bad files are declared : Closes #5387

joeldierkes committed:
Delete the `add_header` script #5389

The `add_header` script output is too volatile for the usage in this
project. The PRs get polluted with the changes made to each python file
header. Merges and other git actions can not be performed easily anymore.

The future of the python file headers and the acknowledgement of open source
contributors need to be discussed and re-evaluated. The job gets disabled
temporarily.

Commit:
Transfers: native multi-hop between transfertools. Closes #5403

Adapt the transfertool argument of the submitter to also accept
a comma-separated list of transfertool names. Most of the needed
functionality to accept such a list was already implemented.
I perform a small refactoring to put the "external_name" of each
transfertool class as a class attribute. This allows for a more
generic code in some parts.

Also adapt the logic which prepares the path submission to a
transfertool. Until now, the submitter was forced to submit the
full path to a single transfertool. This commit changes the
behavior. Now a full path can be splitted into sub-paths, each
of them handled by a different transfertool. In one submitter
iteration, only the first sub-path will be submitted, the rest
will be left in a "queued" state, but protected from being
picked by another submitter thanks to the new "transfer_hops"
table. When the first sub-path will be completed, submission
will resume naturally.

To ensure intermediate hops left in the queued state are picked
by the correct submitter, we now set the "transfertool" field
on intermediate requests. Obviously, the value of this field
can be different from that of the initial request.

Because of the experimental nature of the globus transfertool,
I add a hard-coded protection which ensures that submission is
only done to globus if it's the first element in the submitters
"transfertool" argument.

bari12 committed:
Core: Cleanup of unused fts throttler config file

Commit:
Tests: fix metrics_mock fixture

We recently set the prometheus_multiproc_dir env variable by default
in the env container. As a result, the values written by the mock
are now persisted to disk and hence between test runs when we run
tests in the dev container.
As the pipeline doesn't use this dev container, tests pass without
issues, but the same tests will fail when run locally by a developer.

Fix the fixture by substituting the valueClass with a mutexValue,
ensuring that values aren't persisted between tests.

Also, add the GAUGE and TIMING mock, even if they aren't yet used.

Commit:
Transfers: fix submitter string formatting. Closes #5405

Commit:
Transfers: filter unsupported transfertools early

Instead of doing it near the very end, after path computation,
perform it at the start of request handling.

Commit:
Transfers: per-rse hop_penalty attribute for multihop. Closes #5408

Pass the rse-loaded context through and use it to get rse_info
and attributes.

Remove the un-needed function which it substitutes for the info.

Also fix the global HOP_PENALTY value type. It should be an int.

Commit:
Transfers: disable notifications in globus client

Commit:
Transfers: reduce INFO logging in submitter. Closes #5413

The submitter is, by far, the most log-intensive daemons in our
production.

Move some logs to DEBUG level.

Other logs are combined into one single message to avoid consecutive
messages with little benefit.

joeldierkes committed:
Set the `paramiko` package version to 2.10.3 #5412

`paramiko` package version prior to 2.10.1 has a race condition which could be a
potential security issue.

https://github.com/advisories/GHSA-f8q4-jwww-x3wv

mlassnig committed:
Fix bad Oracle query hints; Fix #5411

ivmfnal committed:
Implemented and tested bulk declaration of bad replicas using lfns, rse, scope
Added api/quarantined_replica.py, removed references to pfns= from calls to declare_bad_replicas()
Make sure UI rse-admin accepts (rse, scope, list-of-LFNs) as arguments to declare multiple replicas as bad (part of #5392)
Reduce chunk_size to 1000, implement unindent() using textwrap.dedent(), fix arguments of declare_bad_file_replicas() call from suspicious_replica_recoverer
Fixed header
Spent another half an hour adding useless annotations
More whitespace beautification
more flake-inspired beautification

Removed unindent function from the rucio-admin - use textwrap.dedent() properly
Allow only list of dicts as the argument for api.quarantined_replica.quarantine_file_replicas()
Removed some trivial method annotations
Added longer description of the dictionary argument arguments desc for QuarantineReplicas in REST

Commit:
Transfers: flag errors coming from tape sources. #5410

cserf committed:
Client : Bug in the upload client for cloud resources : Closes #5354 (#5357)

* Client : Bug in the upload client for cloud resources : Closes #5354

* Fix delete PFNs for cloud services

cfenell committed:
Fix webdav and ssh modules (#5370)

* CA verification option

* WebDAV module bugfixes

* SSH module bugfixes

Commit:
Transfers: remove transfer_limits.py

Functions from this file are not used anywhere and just made us go
on a wrong debugging path while trying throttler.

Commit:
Transfers: fix preparer bulk size argument passing. Closes #5430

cserf committed:
CLI : Problem to print the summary when using rucio add-lifetime-exception : Closes #5427

cserf committed:
Recovery : Too many execution of get_bad_replicas_backlog : Closes #5433

joeldierkes committed:
Add the new add_header script #5455

The old add_header script had multiple problems. This new script provides a new
variant with more features:

- More intuitive user interface
- Easier to add other files support
- Easier to follow the source code

joeldierkes committed:
Add the `add_header` pre-commit step #5455

The `add_header` pre-commit step checks all changed files for correct headers.

joeldierkes committed:
Add the `add_header` job to the CI/CD #5455

The `add_header` job checks the file headers for errors. This is helpfull when a
new file is introduced, since the Lincence information can easily be forgotten.

rizart committed:
Authentication & Authorisation:: Replace deprecated oauth2client dependency Closes #5343 (#5371)

* Authentication & Authorisation: Update GCS url signing #5343

* Authentication & Authorisation: Replace deprecated oauth2 library in requirements #5343

* Authentication & Authorisation: Add myself in the authors #5343

* Authentication & Authorisation: Replace deprecated oauth2 library in server_requirements_table

Commit:
Transfers: handle transfertool assignment failure. Closes #5458

joeldierkes committed:
REST & API: Change the `rses` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

Commit:
Transfers: fallback to transfertool name as protocol for monit. #5432

use the transfertool name if it exists and can be extracted from
the request.

Commit:
Transfers: explicitly set globus protocol for monit. #5432

bari12 committed:
Clients: Disable mmap based checksum calculation; #5462

joeldierkes committed:
Use the present python shebag line instead of python3 #5455

joeldierkes committed:
Fix the file headers #5455

With #5455 the add_header script changed in its behavior. This commit fixes all
headers in all corresponding files.

joeldierkes committed:
REST & API: Change the `requests` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
REST & API: Change the `scopes` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
REST & API: Change the `tmp_dids` api docs to the OpenApi format #4992

The Api documentation is obsolete and not displayed correctly. For this sake we
introduce the OpenApi format for our docs. It is way stricter and a commonly
used format.

joeldierkes committed:
Docker & Kubernetes: Deprecate the `py3` dev docker conainter tag #4198

The `py3` dev docker container tag got deprecated in #5324 due to the
deprecation of Python2 (#5435). A merge of an older PR re-introduced the `py3`
tag.

robbarnsley committed:
Removing drop_database() call after each test.

joeldierkes committed:
Ignore auto generated files in the add_header script #5455

The headers in auto generated files should be ignored, to avoid false positive
hits in the `add_header` check.

joeldierkes committed:
Remove the auto generated header from the autogenerated file #5455

The vcsversion.py gets auto generated. The header was added mistakenly.

joeldierkes committed:
REST & API: Fix the `dids` api docs #4992

The Data Identifier tags had some spelling errors. This lead to the creation of
multiple sidebar-categories in the REST-API documentation.

cserf committed:
Fixes in bad_replicas workflow : Closes #5477

cserf committed:
Catch ORA-00054 in necromancer

cserf committed:
CLI can’t handle input files with more than 1000 datasets : Closes #5479

Commit:
Deletion: rely on raise_exception and default for config_get

simplifies the code

joeldierkes committed:
Add the `count_missing_python_type_annotations` script #4455

With rucio 1.29 python2 got deprecated. This deprecation enables us to introduce
new features used by python3. One of these features is type annotations on the
code itself.

Type hints allow us to run a static type checker like `mypy` and analyze the
code for inconsistencies. Having a strongly typed codebase enables us to avoid
type inconsistency bugs, and it makes the code more readable, thus potentially
shorten the time to add new features.

The plan is to integrate a job in the CI/CD, which reports an error if the
number of missing type annotations is increased by a PR. This is the result of
newly written code, which is not annotated. This forces the user to annotate
their (or potentially some other) code.

This commit introduces the `count_missing_python_type_annotations_rucio`
functions, which uses `flake8` and the flake8 plugin `flake8-annotations` to
count the number of missing type annotations.

`mypy` enables us to check the type annotations for inconsistencies. However, we
can not use `mypy` to count the number of missing type annotations, since it
caps the number of reported errors to 200. We can use `mypy` in a later step to
check the annotations for consistency.

Certain directories are excluded from the type checking:

- The test directory: Type checking the tests is not our goal right now. It
  decreases the time to write test code while providing very little additive.

- The db migration directory: The db migration should focus on the database, not
  the implementation. We could enable this in the future, but it has no priority
  right now.

- The tools directory: Scripts should be written quickly and should not
  necessarily require typed annotations. This can also be changed in the future,
  to enable type-save scripts.

- The bin directory: A strategy is required here on how to handle type
  annotations properly.

joeldierkes committed:
Add the `Checkout ancestor commit` action #4455

Getting the commit from which a branch is branched off of is helpful. The
progress of an implementation in a branch can be measured by comparing some
generated numbers before the feature branch and after. This enables us to do CI
tests based on differences in the committed code to the initial one.

This commit introduces the `Checkout common ancestor` action, which checks out
the most recent ancestor of the current working HEAD and a specified reference.

This implementation can copy files from the edited branch to the most recent
ancestor. This is necessary in case the file got modified and the modified one
is needed in the ancestor commit.

Commit:
Core & Internals: merge rse_attributes.py into rse.py

RseData relied on rsemgr to retrieve the info, but this is equivalent
to calling rse_core.get_rse_protocols with a caching layer around it.

So remove dependency on rsemngr, add a caching wrapper around
get_rse_protocols, and move the RseData into core.rse.

joeldierkes committed:
Check the python type annotation count in the CI #4455

Since we want to introduce python type annotations in our repository, we need a
way to check if new PRs include type annotations.

Since the repository contains a lot of code, we can not add type annotations in
a big PR. We should rather ensure that PRs annotate their new code and gradually
annotate old code.

This commit compared the number of missing type annotations before the PR,
i.e. on the most recent common ancestor commit with the rucio master, and after
the PR. The CI job fails if the number of missing type annotations increases,
since this means that new un-annotated code was introduced. The re-viewers are
responsible for checking the integrity of the type-annotations for now.

Commit:
Deletion: use RseData in reaper

Reduce the number of tuples used for keeping the reaper
execution context by grouping data relative to the RSE into
a single object.

The plan is to further rely on RseData fields in a later
commit.

cserf committed:
Tune permissions for Belle II

Commit:
Deletion: migrate to using RseData.info/attributes in reaper

The commit changes the cache management for rse info. Instead of
retrieving the protocols for all RSEs and storing them in a big
cache entry, each RSEs info is now cached separately. This means
more small cache requests, but also less bursty requests to the
database when cache expires.

Commit:
Deletion: dereference needed_free_space/only_delete_obsolete in loops

Avoid accesses by tuple indexes, which are hard to read.

In one of the loops, only_delete_obsolete is already fetched with a
wrong name and never used. Fix this.

Commit:
Core & Internals: don't fetch rse name in core/get_rse_usage

It's a useless call to the database.

To maintain the 'rse' (name) field on the API level, adapt the
api/get_rse_usage to fill this field by using the data from the
input. Considering that we perform a get_rse_id call just before,
we are sure tha the name corresponds to a valid rse.

cserf committed:
re-evaluate the greedy = 1TB constant

Commit:
Deletion: separate code and cache management in __check_rse_usage

Commit:
Deletion: refactor __check_rse_usage

Stop using the same variable (usage) for different meanings.
Group blocks of code by similar behavior.
Move variable declaration closser to their usage.
Relly on RseData loading function to retrieve needed RSE information.

Commit:
Deletion: move check if rse is staging closer to usage

Commit:
Daemons: move hearbeat handler to daemons/common

Commit:
Deletion: use heartbeat_handler in reaper

Reduce the amount of boilerplate code for handling heartbeats.

Commit:
Deletion: factorize evaluator backlog condition

Avoid having the same complex condition written twice.

Commit:
Deletion: extract deletion worker limiter to separate function

Make the main reaper loop less crowded.

ChristophAmes committed:
Recovery: Implement checksum comparison to suspicious replica recoverer #5334

Before a suspicious file which is the last remaining copy is declared bad, its checksum should be compared to varify that it is corrupt.
A new method is added to /core/replica.py which retrieves the reason for which a replica was declared bad. If the reason is related
to a checksum problem, then the replica is declared bad.

Commit:
Deletion: avoid physical deletion on exceptions in reaper

Commit:
Deletion: separate daemon loop and deletion logic in reaper

This commit slightly changes the current behavior because
some code paths will now sleep for "sleep_time" instead of
a hardcoded 30s.

Commit:
Deletion: de-indent code block in reaper

Commit:
Daemons: move run_daemon to daemons/common

to re-use in other daemons

Commit:
Deletion: use run_daemon in reaper

To reduce the amount of boilerplate code.

Adapted the run_daemon function to have sleep time dependent
on the execution time. This will lower the sleep time in the
conveyor daemons, but will bring the behavior in conformance with
all other daemons (which rely on daemon_sleep() for sleeping).

Commit:
Daemons: force renew heartbeats after waiting for partition

HeartbeatHandler was recently changed to sometimes skip accessing
the database. This makes the logic of partition_wait useless,
because the heartbeats will not be renewed after waiting for
partition.

Commit:
Deletion: remove outer try/except block in reaper run_once

run_daemon already takes care of catching unhandled exceptions
and re-trying the run_once function.

Commit:
Deletion: record MultiGauge rather than prometheus gauge in reaper

Also set directly gauge labels without using a variable

Commit:
Deletion: move info loading closer to usage

Put it in both places where it is used to avoid accessing
un-loaded .info field after potential future refactorings.
RseData ensures that only one database access is done per RSE.

Commit:
Deletion: rework RSE multiplicity logic in reaper. Closes #5198

Reaper had a magic formula which computed how many times it
must re-iterate on the same RSE during a deletion cycle.
As described in the linked ticket, this formula doesn't work
very well when reaper is run with few RSEs. Also, it doesn't
play well with obsolete replica deletion, leading to
starvation of deletion on those RSEs as long as there
is any data to remove on other RSEs.
Also, the formula is not explained anywhere and the reasoning
behind it is not very clear.

Substitute the magic formula with a logic which detects the
RSEs which still have more data to delete and re-iterates
on those RSEs under condition that enough reaper instances
are run in parallel to avoid starvation of other RSEs.

The commit also improves the interaction with run_daemon
wrapper. Reaper will now inform that wrapper that there is
still more work to be done, avoiding to sleep reaper when
there is still data to delete.

Commit:
Monitoring & Logging: reduce gfal2 log level on server side

Commit:
Deletion: pass heartbeat handler into physical deletion loop

the heartbead handler takes care of not updating the database
too frequently.

joeldierkes committed:
Documentation: Change the python header string #5455

The python header contains the creation year of the file. While git does a good
job of tracking file movements and history, it makes more sense to use the
projects creation here.

"CERN" is a bit to short as a name. We will use the full name here.

joeldierkes committed:
Documentation: Add a bash header template #5455

The bash files should have sanitized headers as well.

joeldierkes committed:
Documentation: Fix file headers #5455

cserf committed:
Add Belle II tests

joeldierkes committed:
Clients: Return information on download exception #5250

The different download functions in the `DownloadClient` return lists containing
the files and the states of the corresponding downloads. These states also
include failed states.

If an unknown exception is thrown in the `_download_item` method, the
corresponding file, which should be downloaded, is not present in the output
list. This leads to two problems; both can be solved by adding the file with the
download status `FAILED` to the output list:

- The number of files in the input queue and output queue is different, which
  throws an error. This is easily resolved by setting the file download status
  to `FAILED`. An error is later thrown.

- The client does not print the summary in case there is no items in the output
  queue. By adding the file to the output queue this problem is resolved.

joeldierkes committed:
Clients: Change the download client state to an enum #5250

The client state uses strings to store the state. This has the problem that the
evalutation, if a client state is wrong, is done at runtime. Also, it uses the
exception name as state. This is inconsistent, because all available states
should be well defined.

This commit uses an enum for the download client state. This enables checks at
runtime, since all usable states are defined. The values of the enum ensure the
serialization to strings.

It also changes the usage from an exception name as state to the `FAILING`
state. The exception itself is added in the body already.

Commit:
Core & API: add session to the API level. Fix #5439

Ensure that all API endpoints only initialize/destroy
the session object once. Some API endpoints (did/rules/...)
were already updated in the past. This proved to heavily
reduce the number of spurious rollbacks on the database
side.

cserf committed:
Fix rse_name_generator to match B2 schema

cserf committed:
Introduce new did_name_generator

cserf committed:
Add more tests for DIRAC addfile method and B2 schema validation

Commit:
Core & Internals: ignore some cryptography deprecation warning. #5500

Paramiko.transport imports cryptography.Blowfish, which is deprecated.
Also, python 2 and 3.6 are deprecated in cryptography.
This makes our tests fail due to a warning.
Catch that warning and ignore it with the filter.

cserf committed:
Disable Belle II specific tests from the autotests

rizart committed:
Clients: Force scheme in protocols for delete operations in UploadClient #5507

rizart committed:
Clients: fix typo #5507

Commit:
Transfers: rework preparer workflow. Closes #5502

Rewrite preparer in an imperative way.

The functional paradigm is not used in rucio very much, so
developers don't have experience with it and struggle to
understand the workflow of this code. The benefits are not
justifiable here.

While the big part of the change is unrelated to the linked
issue, also fix a small bug in preparer.

rizart committed:
ESCAPE: Update perm_get_signed_url permission #5513

Commit:
Transfers: tolerate empty request.attributes

Always deserialize it into a dict

joeldierkes committed:
REST & API: Fix the rses api docs #4992

There was a mistake in the indentation of the documentation, which lead to an
error in the generation of the REST API documentation.

cserf committed:
Clean-up unneeded RSEs + Start introducing rse_factory in every tests

Commit:
adding another key for heartbeats permissions.

Commit:
Rules: use run_daemon in judge-cleaner

Commit:
Rules: use run_daemon in judge-evaluator

Commit:
Rules: use run_daemon in judge-injector

Commit:
Rules: use run_daemon in judge-repairer

Commit:
Rules: de-indent code in judges

In each daemon:
 - Remove useless temporary "if True:"
 - Remove else after "if ... return" and de-indent

Commit:
Rules: perform heartbeat.live() in judges sub-loops

Using run_daemon means that the heartbeat refresh frequency and
older_than is computed automatically from "sleep_time". And the
HeartbeatHandler takes care of avoiding to do database calls too
frequently.

However, long intervals between calling heartbeat_handler.live()
may result now in hearbeats getting expired too quickly compared to
the values used previously for older_than (2 hours for the injector
for example). Calling live() inside the inner loop takes care of
this issue as long as each rule doesn't take tens of minutes to
be handled.

Commit:
Tests: migrate test_did.py to pytest

No functional rework at all. Only get rid of unittest.testcase
and use equivalent fixtures.

Commit:
Core & Internals: higher-level primitives for temp table creation

We intend to use 3 kinds of temporary tables in multiple places,
so create primitives for directly instantiating them.

In the future it may allow to easier transit these particular
temporary tables workflow to "global" oracle temp tables.

Commit:
Deletion: use join_from for temporary tables

This construct allows to explicitly specify the table on the left
side of the join. As temporary tables don't have indexes, we want
them to be used on the left side.

I don't know if this is of any use for the optimiser ...

Commit:
Clients: rename DIDType to DID

We already have DIDType in core, with a completely different meaning.
This is confusing.

Commit:
Transfers: sqlalchemy 2 in list_transfer_req_and_source_replicas. #5247

cserf committed:
Improvements in the Lifetime Model pages : Closes #1923

Commit:
Metadata: use run_daemon in abacus. #5443

Commit:
Metadata: fix code indentation in abacus daemons

Commit:
Core & Internals: core/did queries in sqlalchemy2 syntax. #5247

with_hint is not implemented for oracle on update(), so migrate to
using prefix_with for one query.

This is a preparatory work for temporary table support in core/did.

cserf committed:
Add ATLAS specific tests back

Fix headers

Commit:
Core & Internals: remove redundant and_ in core/did queries

there is an implicit and in where()

joeldierkes committed:
Documentation: Fix the file headers #5455

The `lib/rucio/daemons/common.py` file header was poorly formatted. This came
due to the creation of the file after the PR which fixes all the headers. The
header was on a previous version, since only a rebase was run, not the
`add_header` script itself.  A simple `tools/add_headers` run fixes this issue.

joeldierkes committed:
Documentation: use files known to git in the `add_header` script #5455

The `add_header` script should only consider files known to git. This preserves
the integrity of user and auto generated files ignored by the VS (like in the
`.venv`folder) and saves time. The previously used simple algorithm to
determine, if a file should be ignored, was to bold for that.

Commit:
Tests: fix some grammar mistakes in test_reaper comments

Commit:
Deletion: add support for oidc in reaper. Closes #5538

This is a very basic support, mostly replicated from how conveyor
does the oidc authentication.

joeldierkes committed:
Documentation: Fix wrongly formatted param doc comments #5547

Some function doc comments do not follow the sphinx documentation standard:

- Some are missing the double-point after the parameter name, e.g.
  `:param rse The RSE...`.

- Some have the double-point directly after `:param`, e.g.
  `:param: rse The RSE...`.

Both should be `:param rse: The RSE...`.

bari12 commented on .gitignore:
```diff
@@ -65,4 +65,3 @@ tools/test.file.1G
 
 # autotest logs/data
 /.autotest
-
```
Seems an unnecessary change? 😄

joeldierkes replied:
I think he activated the `pre-commit`s for the project. They are fixing the file endings for me as well. We can deactivate that, does not seem that important.
https://github.com/rucio/rucio/blob/a4db9e66ba5a175a721f75bb59b0761c92e8db5a/.pre-commit-config.yaml#L5

bari12 commented on .gitignore:
```diff
@@ -65,4 +65,3 @@ tools/test.file.1G
 
 # autotest logs/data
 /.autotest
-
```
Seems an unnecessary change? 😄

joeldierkes replied:
This is fixed now, a rebase should solve the problem.

bari12 commented on bin/rucio-admin:
```diff
@@ -1041,15 +1066,81 @@ def delete_account_attribute(args):
+    chunk = []
+
+    # send requests in chunks
+    chunk_size = 10000
```
Did you test that 10k works? Seems a lot in terms of header sizes?

bari12 replied:
Did you test that 10k works? Seems a lot in terms of header sizes?

bari12 commented on bin/rucio-admin:
```diff
@@ -1041,15 +1066,81 @@ def delete_account_attribute(args):
 
 
+@exception_handler
+def quarantine_replicas(args):
```
Add function description

bari12 replied:
Add function description

bari12 commented on bin/rucio-admin:
```diff
@@ -1041,15 +1066,81 @@ def delete_account_attribute(args):
+
+def __declare_bad_file_replicas_by_lfns(args):
+    """
+
```
Remove spaces around description

bari12 replied:
Remove spaces around description

bari12 commented on lib/rucio/api/quarantined_replica.py:
```diff
@@ -0,0 +1,72 @@
+# limitations under the License.
+#
+# Authors:
+# - Igor Mandrichenko <ivm@fnal.gov>, 2022
```
Please re-run all these against the new `add_header` script

bari12 replied:
Please re-run all these against the new `add_header` script

bari12 commented on lib/rucio/web/ui/static/webui_version:
```diff
@@ -1 +1 @@
-1.28.0
\ No newline at end of file
+1.28.0
```
Please remove that change

bari12 replied:
Please remove that change

bari12 reviewed (changes requested):
Hi Igor, sorry for the late review. I think in general it is ok, I marked some minor changes.
You also need to rebase since there were some other changes in the meanwhile.

@cserf please have a look as well

Commit:
Deletion: pause reaper if greedyDeletion but nothing to do. Closes #5549

Greedy deletion is a special case when we don't have a specific
amount of bytes we want to remove. So chunk_size is variable.
Howether, when there is absolutely no work to do, we still have
to stop the deletion on those RSES, or reaper will hammer the
database without any reason.

bari12 commented on lib/rucio/web/rest/flaskapi/v1/replicas.py:
```diff
@@ -892,6 +894,46 @@ def post(self):
 
+class QuarantineReplicas(ErrorHandlingMethodView):
+
+    def post(self):
```
Before I forget, the documentation for the REST endpoints needs to be written a little bit different now. See
https://github.com/rucio/rucio/blob/2aedf7f83cdb88f5f47a81a6ee29588afad13e6b/lib/rucio/web/rest/flaskapi/v1/dids.py#L373
The reason for this is so it is parsable by our now rest endpoint doc: http://rucio.cern.ch/documentation/html/rest_api_doc.html
Please check with @joeldierkes (on slack?) if you have any questions on it.

joeldierkes replied:
Yes, feel free to write me if you have any questions. Other files like [rses.py](https://github.com/rucio/rucio/blob/master/lib/rucio/web/rest/flaskapi/v1/rses.py) already implement these style changes.

Commit:
Core & Internals: disallow circular container attachment. Closes #5523

joeldierkes committed:
Consistency checks: Fix the project structure #5113

Python projects use `__init__.py` files to mark directories as part of a
`regular package` [1]. The content of the regular package can get importet by
python via the `import` keyword.

The `storage/consistency` folders both do not contain such files, which makes
them inaccessible to global imports.

[1] https://docs.python.org/3/reference/import.html#regular-packages

cserf committed:
Many database queries caused by necromancer : Closes #5545

Commit:
Traces: reduce log verbosity in tracer-kronos. Closes #5556

Substitute the log with a counter to record the metric into
prometheus and statsd.

dchristidis commented on bin/rucio-admin:
```diff
@@ -96,6 +96,31 @@ logger = logging.getLogger("user")
+        out_lines.append(add_indent + line)
+    return "\n".join(out_lines)
+
+
```
Maybe I misunderstood your motivation behind this, but I believe the use of `textwrap.dedent()` to be superior.

joeldierkes committed:
Testing: Disable GitHub Action Scheduled Autotest Runs in Forks #5560

The GitHub actions scheduled autotest runs are also executed in the personal
forks. This is unnecessary, since they are already executed in the main
repository.

This commit only runs the scheduled autotests in the main github actions.

Commit:
Transfers: skip transfer if integrity error on source sync. Closes #5565

Rework submitter workflow. Fetching the requests and path creation
is not done in the same transaction anymore. This way, when submitter
fails to create sources, there is no need to abandon all the db_bulk
anymore. We can skip the offending request and continue with the next
ones.

This allows to recover from foreign key violation on the source table
and skip those transfers.

joeldierkes committed:
Testing: Fix wrong bash file headers #5483

Commit:
Transfers: de-indent code block

joeldierkes committed:
Clients: Remove the DQ2 Client #5584

The DQ2 client was introduced to maintain backward compatibility to the
previous CMS system. This backward compatibility however is not required
anymore, since the system is not used anymore.

Commit:
Transfers: always skip ACTIVE transfers in receiver. Closes #5335

the job_m_replica flag is for multi-source transfers, so the condition
which I remove makes no sense. We don't want to transition the state
of a multi-source transfer which is still in an ACTIVE state.

While the commit fixes the potential bug from the linked ticket, the
bug didn't occur in production because of a bug in FTS: job_m_replica
was always returned False by fts, even for multi-source transfers.
The FTS bug was already fixed and will be live soon (with the release
of FTS v3.12.0).

rizart committed:
Protocols : update SCHEME_MAP for cs3s #5586

joeldierkes committed:
Testing: Disable non-determined rucio version in autotests #5572

The `autotest` docker images get build on a daily basis. There are multiple
images build with different configurations (different OS, Python version, Rucio
version). While the OS and the Python version is encoded in the image tag
(e.g. `ghcr.io/rucio/rucio/rucio-autotest:fedora33-python3.9`), the Rucio
version is not.

We are building four different classes of images every day with various
configurations: `Release-1.23-LTS`, `Release-1.26-LTS`, `Release-1.28`,
`master`. Since the tag does not contain the Rucio version, the image available
online corresponds to the last image uploaded. So the image is up-to-date if the
`master` finished last, but not if some other configuration does.

This is **not** a problem in the `autotest` workflow run, since the image gets
build every time to make up for changes. This, however, is a problem if we have
a process which needs an up-to-date (tracking master) container
(e.g. rucio/documentation#115), since the Rucio and dependency version in the
container is non-deterministic.

There are two possibilities to proceed here:

1. Include the rucio version in the tag. This way, we have up-to-date LTS
  containers as well as one tracking the master. This however rebuilds the
  containers every night, even if the LTS ones are not really used.

2. Only keep the latest project version and don't include the other ones. This
  saves time on a daily basis and only forces a build of the other versions if a
  build is needed. We also don't have to change the `autotest` code.

We decided for the 2. option and skip the build of the other containers.

joeldierkes committed:
Documentation: Change the file header of `.py.mako` files #5597

`.py.mako` files are template files used to generate python files. The
file headers should be treated the same as python ones.

rizart committed:
Database: Fix alembic offline mode migration logic #5488 (#5490)

* Database: add alembic offline template #5488

* Database: modify alembic migration script for offline mode #5488

* Database: remove whitespace from blank line #5488

* Database: rename alembic offline .ini template #5488

* Database: update .ini headers #5488

* Database: fix headers for alembic .ini files #5488

joeldierkes committed:
Clients: Fix various rucio bin's help text formatting #5537

By default, `argparse` warps the _description_ and _epilog_ of the help
text. Text, which should be displayed as it is in the code, also gets
wrapped. This is a problem, since it makes it harder to read and
basically impossible to understand code examples.

E.g.:
```bash
$ rucio-abacus-rse --help
usage: rucio-abacus-rse [-h] [--run-once] [--threads THREADS]
                        [--enable-history] [--sleep-time SLEEP_TIME]

The Abacus-RSE daemon is responsible for updating RSE usages. It checks if
there are new entries in the UpdatedRSECounter table and updates the RSE
counter in the RSECounter table by adding or substrating the amount of files
and the size.

optional arguments:
  -h, --help            show this help message and exit
  --run-once            One iteration only
  --threads THREADS     Concurrency control: total number of threads on this
                        process
  --enable-history      Record RSE usage into history table every hour.
  --sleep-time SLEEP_TIME
                        Concurrency control: thread sleep time after each
                        chunk of work

Upload a file to your RSE:: $ rucio upload --rse MOCK --scope mock
filename.txt Check RSE usage:: $ rucio list-rse-usage MOCK USAGE: ------
files: 0 used: 0.000 B rse: MOCK updated_at: 2018-11-30 14:28:34 source: rucio
------ Run the daemon:: $ rucio-abacus-rse --run-once Check RSE usage again::
$ rucio list-rse-usage MOCK USAGE: ------ files: 1 used: 213.481 kB rse: MOCK
updated_at: 2018-12-03 08:58:33 source: rucio ------
```

This commit changes `argparse`s behaviour, it displays the lines
correctly now.

```bash
$ rucio-abacus-rse --help
usage: rucio-abacus-rse [-h] [--run-once] [--threads THREADS]
                        [--enable-history] [--sleep-time SLEEP_TIME]

The Abacus-RSE daemon is responsible for updating RSE usages. It checks if there are new entries in the UpdatedRSECounter table and updates the RSE counter in the RSECounter table by adding or substrating the amount of files and the size.

optional arguments:
  -h, --help            show this help message and exit
  --run-once            One iteration only
  --threads THREADS     Concurrency control: total number of threads on this
                        process
  --enable-history      Record RSE usage into history table every hour.
  --sleep-time SLEEP_TIME
                        Concurrency control: thread sleep time after each
                        chunk of work

Upload a file to your RSE::

  $ rucio upload --rse MOCK --scope mock filename.txt

Check RSE usage::

  $ rucio list-rse-usage MOCK
  USAGE:
  ------
     files: 0
     used: 0.000 B
     rse: MOCK
     updated_at: 2018-11-30 14:28:34
     source: rucio
  ------

Run the daemon::

  $ rucio-abacus-rse --run-once

Check RSE usage again::

  $ rucio list-rse-usage MOCK
  USAGE:
  ------
      files: 1
      used: 213.481 kB
      rse: MOCK
      updated_at: 2018-12-03 08:58:33
      source: rucio
  ------
```

rizart committed:
Metadata: Add additional operators in metadata filter engine for Oracle #5503  (#5567)

* Metadata: Add additional operators in metadata filter engine for Oracle #5503

* Metadata: fix column name #5503

* Metadata: Include Oracle in tests #5503

* Metadata: Remove nonexistent-key results for json metadata #5503

* Metadata: Adjust tests for nonexistent-key results #5503

* Metadata: Adjust tests for nonexistent-key results #5503

* Metadata: fix did result number in comment #5503

* Metadata: fix typo #5503

* Metadata: Fix wrong DataIdentifierNotFound exception #4512 #5503

bari12 committed:
Revert "Core & Internals: remove redundant and_ in core/did queries"

This reverts commit 19c8d80e2ea58066a2d308e35a564311e0b650ec.

bari12 committed:
Revert "Core & Internals: core/did queries in sqlalchemy2 syntax. #5247"

This reverts commit bbf7cc11a8a0a0fb3f6feffb9bc69b9e5a39f14f.

bari12 committed:
Database: Add query hint to second code path for list_dids

joeldierkes commented on lib/rucio/web/rest/flaskapi/v1/replicas.py:
```diff
@@ -883,6 +884,57 @@ def post(self):
+          404:
+            description: RSENotFound
+          503:
+            description: ResourceTemporaryUnavailable
```
Good job, however I'd like this part to be a bit more specific. The API documentation should be verbose on its own, since we should not force users to look into the code to figure out what the input and output of the endpoint is.

Something like
```text
requestBody:
  content:
    application/json:
      schema:
        type: object
        required:
          - rse
          - rse_id
          - replicas
        properties:
          rse:
            description: The name of the rse.
            type: string
          rse_id:
            description: The id of the rse.
            type: string
          replicas:
            description: The replicas to quarantine.
            type: array
            items:
              type: object
              properties:
                scope:
                  description: The scope of the replica.
                  type: string
                name:
                  description: The name of the replica.
                  type: string
                path:
                  description: The path to the replica.
                  type: string
```

joeldierkes commented on lib/rucio/api/quarantined_replica.py:
```diff
@@ -0,0 +1,69 @@
+    """
+    Quarantine replicas.
+
+    :param replicas: Either a list of paths (string) or a list of replica infos: {'scope': <scope> (optional), 'name': <name> (optional), 'path':<path>}.
```
Why do we allow a list of either strings or dicts here?
Isn't a list of dicts with some optional values enough?

If it is really necessary, I'd suggest splitting this into two functions:
```python
quarantine_file_replicas_string_list(replicas, ...):
    replicas = [{'path': p} for p in replicas]
    quarantine_file_replicas(replicas, ...)

quarantine_file_replicas(replicas, ...): 
```
But only if it is **really** needed!

joeldierkes commented on bin/rucio-admin:
```diff
@@ -67,6 +68,17 @@ logger = logging.getLogger("user")
 tablefmt = 'psql'
 
 
+def unindent(text: object, indent: object = 0) -> object:
```
The name `unindent` is misleading, since the actual unindentation starts from the second line. Maybe `unindent_without_first_line` or `unindent_after_first_line` or `unindent_from_second_line`?

dchristidis replied:
My original comment still stands: the need for this function is unclear. The `textwrap.dedent()` function from the standard library can be used directly by having the first line of the triple-quoted string be a backslash.

joeldierkes replied:
Yes I agree. The example for the `dedent` function is clear: https://docs.python.org/3/library/textwrap.html#textwrap.dedent
The first line only has to be a backslash, and the next line should be indented as well. I'm also in favor of this change. :+1:

bari12 reviewed (commented):
Please rebase this, there are some conflicts;
Also please remove the `object` type annotations. I'll still merge it without them (even though the CI test might fire) - but these ones might mislead people in adding their own annotations to their PRs.
Otherwise this is fine; I will add it either to 1.29.0 or 1.29.1

ivmfnal committed:
removed object as type annotation

ivmfnal commented:
Reopened as 5773
